### ISSUE TITLE
pubsub: pstest with Receive, 1s delay

### ISSUE THREAD

**Issue by JeanMertz** [GoogleCloudPlatform/google-cloud-go#875]
I've got a test that starts the `pstest` server, instantiates a client, creates a topic, a subscription, and publishes a message to that topic. At this point, I'm `0.009` seconds into running the test (this is with a closed `PublishResult.Ready`, so the message should already have been delivered to the server at this point).

Now I want to read that message, so I start `subscription.Receive` and wait for the callback. At this point, the callback takes a while, and the next point in time I can measure, `1.011` seconds have passed.

Digging a bit further, it seems like the delay actually occurs here:

https://github.com/GoogleCloudPlatform/google-cloud-go/blob/90bed45ceade066ca2d7c9b9954649f9822e46c6/pubsub/subscription.go#L475

In other words, the delay is in the actual receiving of the message from the server, even if the message is already ready for delivery.

Is there any way to speed up message delivery using the test server to speed up a test-suite with many tests and multiple test-servers being used?

I looked into `Subscription.ReceiveSettings` and `Topic.PublishSettings`, but it doesn't look like any of them influence this delay.

**Comment by jba** [comment]
There's a polling loop in the fake server, with a sleep of 1 second: https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/pubsub/pstest/fake.go#L501.

I make it shorter in https://code-review.googlesource.com/c/gocloud/+/22470.

**Comment by JeanMertz** [comment]
Awesome. I totally missed that one. Any reason to have it still be 100ms? It's a fake server, so resource utilisation shouldn't be a concern right? 100ms is a 10x speedup, but 10ms would be even better.

Either way, thanks for this improvement 👍

**Comment by jba** [comment]
Changed to 10ms.

**Comment by jonathangjertsen** [comment]
There is still a delay here and I'm running into it in a test case where I publish lots of messages concurrently. All streams quickly become busy, so this break happens:

https://github.com/googleapis/google-cloud-go/blob/a1a30091f329a3b22e8510e5089c0d95fc632e9d/pubsub/pstest/fake.go#L1234

causing the server to wait 10ms before trying again:

https://github.com/googleapis/google-cloud-go/blob/a1a30091f329a3b22e8510e5089c0d95fc632e9d/pubsub/pstest/fake.go#L959

I don't expect similar issues to occur in a real Pub/Sub system using the streaming pull api. Instead of waiting a fixed 10ms, shouldn't it try to deliver again as soon as a stream becomes ready?

A workaround I found is to increase MaxGoroutines to some large number in the test.
